Add LevelDB blockDB operational mode to replace sequential block files

Greg-Griffith opened the pull request:
For full information about the goal of this PR please reference: https://bitco.in/forum/threads/devpool-work-replace-block-files-with-key-value-db.8532/

#### Note: when testing, you will want to run this on a unique datadir path.

To do:
-python test to validate blockdb

Commit:
add leveldb for blocks

add the base code for a leveldb to hold block data.
move read/write/pruning block functions from main.cpp to blockdb/blockdb_sequential
make leveldb equivalents to the sequential read/write/pruning functions
make a wrapper to wrap storage method specific read/write/prune functions to prevent adding adding decision logic for the functions whenever they are used.
add arg to determine storage mode
set default storage mode to leveldb
add a pointer to the block leveldb called pblockfull
add function to CBlockTreeDB to find information on a specific index by its hash
add different best block for sequential files and for leveldb

kyuupichan commented:
Hi Greg, one concern I would have about this is leveldb moving data around between levels.  This tends to mean a lot more disk I/O than would happen otherwise (the so called write amplification issue).

ptschip commented:
I believe there is a setting somewhere that will make level db not cache
data when you read from disk, however, you make a good point that there
may be a performance impact and that needs to be quantified at some
point during testing.
>
> Hi Greg, one concern I would have about this is leveldb moving data
> around between levels. This tends to mean a lot more disk I/O than
> would happen otherwise (the so called write amplification issue).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/989#issuecomment-370139034>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0CY4xldQgg_bgIcuvYsbwSxRx77cks5tanidgaJpZM4SaqAE>.
>

ptschip commented:
on the other hand, it may prove very beneficial to have a small in
memory level db cache that can hold perhaps perhaps 50 or 100MB...this
would be great when handling blocks that arrive out of order and then
don't have to be read from disk; during initial sync this could be a big
plus as blocks often will arrive out of order and quite a few of them
can...at any rate all this would need to be tested out and tuned for the
best performance.


On 03/03/2018 3:03 AM, Neil wrote:
>
> Hi Greg, one concern I would have about this is leveldb moving data
> around between levels. This tends to mean a lot more disk I/O than
> would happen otherwise (the so called write amplification issue).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/989#issuecomment-370139034>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0CY4xldQgg_bgIcuvYsbwSxRx77cks5tanidgaJpZM4SaqAE>.
>

kyuupichan commented:
I would hope the OS would cache as well as leveldb.

What's the problem with using the filesystem for blocks?  Indexing of those could be in leveldb. If removing blocks because of pruning you'd just need some very occasional compaction of files or suchlike.  Most of the time the files aren't going to change once done. In ElectrumX transaction hashes of all transactions are stored linearly on disk and it works well.

Greg-Griffith commented:
The current system works by storing CBlockIndex data and block file information in leveldb. The blocks are stored in sequential order on disk in the block files. So i think what you just suggested is the system currently in place @kyuupichan

gandrewstone commented:
We are writing 1 block in 10 minutes so that's not going to be a perf issue.  IBD may be an issue, we will need to see.  I feel that DIY pruning choices are essentially reinventing a database. if leveldb turns out to suck at this task surely we should be able to find a DB optimized for append and read, allowing rare deletes.

Greg-Griffith commented:
switching from sequential blocks to levelDB now works without needing to reindex, going the other way still requires a reindex. 

when switching to levelDB the debug.log is flooded with a lot of messages. which needs to be fixed. 

dual db mode is still untested and more than likely doesnt work.

ptschip commented:
I'll give this a run tomorrow but I see you're still getting a lot of test failures on travis?

ptschip commented:
@Greg-Griffith Also was wondering if you could rebase this on top of the current dev. This looks like it doesn't have the code for pruning slow peers and may be the reason you're getting some slow IBD.

ptschip commented:
@Greg-Griffith I just did quick short run.  IBD is painfully slow even in blockdbtype=0,  I think you should definitely rebase on top of the current dev and we should see some better performance.

ptschip commented:
@Greg-Griffith ok got it work well enough...banned the slow peers.  Here are my quick results for the first part of the chain.  Performance with dbtype=0 is normal, we do about 60,000 blocks for the first minute...with dbtype=1 we do only 2000 blocks per minute.  Quite a big difference.  I wonder whether you are flushing the db on every block you get rather than maintaining a small cache of 10 or 20 mb before flushing?

ptschip commented:
@Greg-Griffith just got a really big boost to performance by using some levedb option:      iteroptions.fill_cache = true;  

Was really humming along and then crashed at about 160,000 blocks, but we were only running for about 4 minutes...so we're almost as fast as regular IBD.

ptschip commented on src/init.cpp:
```diff
@@ -991,9 +994,11 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
                 uiInterface.InitMessage(_("Opening Block database..."));
                 pblocktree = new CBlockTreeDB(nBlockTreeDBCache, false, fReindex);
+                pblockfull = new CFullBlockDB(nBlockTreeDBCache, false, false);
                 uiInterface.InitMessage(_("Opening UTXO database..."));
```
we shouldn't be using nBlockTreeDBCache here...we need to calculate a new setting.  It's a bit of work because our dynamic cache settings.

ptschip commented on src/allowed_args.cpp:
```diff
@@ -230,6 +231,8 @@ static void addGeneralOptions(AllowedArgs &allowedArgs, HelpMessageMode mode)
             strprintf(_("Whether to operate in a blocks only mode (default: %u)"), DEFAULT_BLOCKSONLY))
+        .addArg("blockdbtype=<n>", requiredInt,
+            strprintf(_("Which method to store blocks on disk (default: %u) 1 = sequential files, 2 = leveldb, 3 = both")), DEFAULT_BLOCK_DB_MODE)
         .addArg("checkblocks=<n>", requiredInt,
```
so if we're going to support both simultaneously, then what are the rules around a chain rollback or reorg?  which set of blocks are we expected to use?  Isn't this possibly going to cause confusion for the devs in the future; maybe we should just do one or the other but not both?  food for thought.

ptschip committed:
Prevent sequential block files from being created

Block files were being created during IBD when only in LevelDB mode
due to the FindBlockPos() which should not be executed when in
LevelDB mode.

ptschip commented:
needs a rebase

Commit:
add method to determine if we can sync storage without network

rename blockdb_sequential to sequential_files

rename blockdb_wrapper to wrapper

rename blockdb_leveldb to blockdb

Commit:
add wrapper for undo data storage

Commit:
add wrapper to getbestblock

add wrapper to getbestblock
add read and write for bestblock for both sequential files and db storage

minor compilation fixes

Commit:
fix improper undodb creation, fix undo data write logic

- undodb was never properly being assigned, its location was being assigned to the block db.
- we change logic to write undo info regardless of db mode. we should only find the undo pos in sequential mode however since db mode doesnt use a pos.

getbestblock to correct method, fix bad include

undo data now written for db when blocks sync

some improvements on copying data

much more reliable way of syncing files -> db

remove exist checks, only works without them

initialize to c++11 nullptr not to NULL

sickpig commented:
need a rebase

sickpig commented on src/blockdb/blockdb.cpp:
```diff
@@ -0,0 +1,107 @@
+#include "blockdb.h"
+#include "hash.h"
+
+CBlockDB *pblockdb = NULL;
```
nit: use c++11 nullptr as initialization value.

```c++
CBlockDB *pblockdb = nullptr
```

sickpig commented on src/blockdb/blockdb.cpp:
```diff
@@ -0,0 +1,107 @@
+bool CBlockDB::WriteBatchSync(const std::vector<CBlock> &blocks)
+{
+    CDBBatch batch(*this);
+    for (std::vector<CBlock>::const_iterator it = blocks.begin(); it != blocks.end(); it++)
```
nit: use c++11 range `for` statement 

```c++
for (const auto &it : blocks) 
```

(this applies also to all the rest of the code)

sickpig commented on src/blockdb/blockdb.cpp:
```diff
@@ -0,0 +1,107 @@
+{
+    std::vector<uint256> hashesToPrune;
+    /// just remove the to be pruned blocks here in the case of leveldb storage
+    boost::scoped_ptr<CDBIterator> pcursor(pblockdb->NewIterator());
```
nit: use c++11 `unique_ptr`

sickpig commented on src/blockdb/blockdb.h:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    template <typename Stream>
+    void Unserialize(Stream &s)
```
nit: `Deserialize`

Greg-Griffith replied:
using unserialize keeps it consistent with the rest of the code. see serialize.h for examples

sickpig replied:
fair enough. It was just that it sounded wrong to me :smile:

sickpig commented on src/blockdb/sequential_files.cpp:
```diff
@@ -0,0 +1,306 @@
+// Copyright (c) 2009-2015 The Bitcoin Core developers
+// Copyright (c) 2015-2018 The Bitcoin Unlimited developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
this is a new file no need to include any copyright header earlier than 2018.

Greg-Griffith replied:
I was unsure what to do about this. some of the functions in this file were moved from main.cpp which has the copyright for core. if I copy and pasted their code does credit need to be maintained or does it simply being a new file overwrite this need?

sickpig replied:
IANAL I did not consider that code move was involved. Taking that into consideration I guess you did the right thing.

sickpig commented on src/blockdb/sequential_files.cpp:
```diff
@@ -0,0 +1,306 @@
+        CDiskBlockPos pos(*it, 0);
+        fs::remove(GetBlockPosFilename(pos, "blk"));
+        fs::remove(GetBlockPosFilename(pos, "rev"));
+        LOGA("Prune: %s deleted blk/rev (%05u)\n", __func__, *it);
```
wonder if we could bind this log to some existing categories (or create a new one) rather than log it always.

edit: saw that a `PRUNE` category exists

sickpig commented on src/blockdb/sequential_files.cpp:
```diff
@@ -0,0 +1,306 @@
+uint64_t CalculateCurrentUsage()
+{
+    uint64_t retval = 0;
+    BOOST_FOREACH (const CBlockFileInfo &file, vinfoBlockFile)
```
use c++11 range `for` statement

(we are trying to use get away from boost where possible)

sickpig commented on src/blockdb/sequential_files.h:
```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2015 The Bitcoin Core developers
+// Copyright (c) 2015-2018 The Bitcoin Unlimited developers
+// Distributed under the MIT software license, see the accompanying
```
new file, no need to ass any copyright header earlier than 2018

sickpig commented on src/blockdb/wrapper.cpp:
```diff
@@ -0,0 +1,545 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2015 The Bitcoin Core developers
```
ditto

sickpig commented on src/blockdb/wrapper.cpp:
```diff
@@ -0,0 +1,545 @@
+
+bool DetermineStorageSync()
+{
+    uint256 bestHashSeq = pcoinsdbview->GetBestBlockSeq();
```
use `LOG(SOMECATEGORY,"..")` rather than `LOGA` in this function

awemany replied:
Also maybe something more descriptive than `"check1"` etc.?

sickpig commented on src/blockdb/wrapper.cpp:
```diff
@@ -0,0 +1,545 @@
+            if(it == mapBlockIndex.end())
+            {
+                LOGA("something is very wrong somewhere \n");
+                assert(false);
```
why no using `AbortNode` rather than asserting to false? 
(contrary to what usually happen, in bitcoin land, asserting trigs a crash also when running in normal mode)

same question for all the other asserts in this and the next function

Greg-Griffith replied:
whenever this failure occurred it would assert false later on an expression check anyway. this worked out to be a case of assert false now or wait to do it later. I chose to add one to just do it now.

sickpig replied:
but didn't `AbortNode` just shutdown the node immediatly?

sickpig commented on src/blockdb/wrapper.h:
```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2015 The Bitcoin Core developers
+// Copyright (c) 2015-2018 The Bitcoin Unlimited developers
+// Distributed under the MIT software license, see the accompanying
```
no need to have any headers earlier than 2018

sickpig commented on src/init.cpp:
```diff
@@ -937,36 +943,55 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
     // ********************************************************* Step 6: load block chain
 
     fReindex = GetBoolArg("-reindex", DEFAULT_REINDEX);
+    int64_t requested_block_mode = GetArg("-blockdbtype", 1);
```
use `DEFAULT_BLOCK_DB_MODE` as second argument of `GetArg` so you wouldn't need to bother to special case for the default value (next if)

sickpig commented on src/main.cpp:
```diff
@@ -4152,7 +3748,37 @@ bool static LoadBlockIndexDB()
+
+    if(!fHavePruned)
+    {
+        LOGA("blocks arent pruned, trying to sync between storage methods\n");
```
use LOG(PRUNE,"...")

sickpig commented:
need rebase

Commit:
improve syncstorage for sequential file mode

ptschip commented:
@Greg-Griffith I wonder whether FlushStateToDisk() was better left in main.cpp rather than putting it into the wrapper.cpp.  FlushStateToDisk() is not just used for flushing block state but also for the utxo and possibly other things in the future.

Greg-Griffith commented:
@ptschip what model is your ssd? I test on 850 EVO (SSD) and WD Red(HDD). I have not come across that problem with either of those drives.

ptschip commented:
On another topic I do get these "Failed to connect to best block" occasionally on restart.  I turned of PV and commented out the shutdown code relating to the validation threads but the problem still persists...however if I run in non-blocksdb I don't have the issue.

ptschip commented:
This is what i see in the log file regarding the best block issue:

2018-06-15 01:34:34.270990 height of best seq block = 0 
2018-06-15 01:34:34.270990 height of best db block = 90454 
2018-06-15 01:34:34.318458 ERROR: ConnectBlock(): tried to overwrite transaction

the best db block is correct but the best seq block is wrong ?

ptschip commented:
never mind, looks like we're behind by a couple of blocks....hmmm.

ptschip commented:
OK I got to the bottom of it and happily it also fixes the crash bug...I'll put together a PR to your repo tomorrow.  Basically the bestblock hash was not being updated at the right place.

ptschip committed:
Set the best block hash in ConnectBlock() and DisconnectBlock()

We have to set the best block hash at these places. If we do it
later after ActivateBestChain then we may end up a few blocks
behind because ActivateBestChain can connect many blocks at at
time and then have its thread interrupted which means the best
block hash will be incorrect. It's better to update the block
hash directly after connecting or disconnecting a block.

use readwrite over varint due to issue 1144

return false if we match other storage method

pos system is not used by blockdb

write db_best_block when we would write db_best_block

remove writebestblockdb as view now handles it

blockvalue serialization fixes

use pos -1 as null to determine if data is in files or blockdb

if nStatus says we have the data but pos is -1 then it is in blockdb, if pos is >=0 then it is in files or both in files and on blockdb.

the negative unsigned int is intentional.

new logic for writing files from blockdb

cleanup code determining besthash on startup

cleanup BLOCK_DB_MODE assignment on startup

cleanup assignment logic. changed argument name to improve use clarity

Fix 2's complement issue. Requires Reindex

skip negative index files, they dont exist

add cblockindex fields and new disk storage

use additional fields in cblockindex and a new CDbBlockIndex class to store that additional data. we are able to import from old CDiskBlockIndex and this keeps better compatability with the nPoS nFile system that the other bch implementations use. we should be able to swap the client with and ABC node and still be able to startup without any issues.

restore remaining npos fields to original value

trigger import on startup, importing is a one time function

Greg-Griffith commented:
yeah. as soon as you said that I knew I should have put it at the point where tip actually changes like the older systems had it

Commit:
use a seperate index system for blockdb

fix compiler warning for order of opetations

reset default nFile value for diskpos

Commit:
implement pblocktreeother

 to read data from the storage method we are currently not using during the initial syncing process. it is not used at all once blockindex is initialized. this is a lot easier than navigating locks trying to open and close this the multiple places we need it

restore varint to readwrite for cdiskblockindex

fix issue with uninitalized pointers

use CDiskBlockIndex instead of CBlockIndex for findblockindex due to the way hashing for cblockindex works

remove setNull for pos for blockdb

causes file of -1 when it should be a value > 0

Commit:
adjust logic for genesis and adding data

adjust genesis logic for sequential file sync

fix spelling, add some debug info

fix use of wrong pointer

we should be using pindexNew instead of it because it == mapblockindex.end() in this conditional

set npos flags for blockdb

unwrite data if we cant copy the data

return the sorted CDiskBlockIndexes instead of just their hashes

remove back check for sync, add on-the-fly repair

Greg-Griffith commented:
Explanation on the last block of commits which is a merge from blockdb2 branch on my repo:

In the current blockdb code I changed the way CBlockIndex worked to account for a second file system but some of the required changes break compatibility with other BCH nodes such as ABC and XT. The end result being a required reindex any time a user were to switch between BU and ABC or XT. blockdb2 incorporates its own index system for blockdb to retain compatibility with the other nodes. We sync between storage types when starting the BU node to prevent using the network to sync data we already have elsewhere on disk

Commit:
sickpig nits and other cleanup

healing the sickpig

ptschip committed:
Create a database key that will sort by time for the block files

By prefixing the block hash with the blocktime plus a ":" separator
the records will now be sorted by time on disk. This has a large
perforance benefit in reducing the frequency and scope of database
compactions.  The only files that now undergo compaction are the most
recent one created, rather than any and all files as was the previous
case.

ptschip commented:
needs a rebase

ptschip committed:
Add new parameters to GetCacheConfiguration

We need to be able to auto size the level db block and undo caches
in the same way we do for the block index and coins cache.

ptschip committed:
Calculate the return value of the new parameters and apply them

ptschip committed:
Rename GetOptions() to GetDefaultOptions()

ptschip committed:
Create COverrideOptions struct and use it

This allows us the ability to modify the defalt leveldb options that
we apply to every level db instance.  Some db's, such as the one used
for holding blocks, may want to have different options for things like
file size, block size etc.

Override max file size for blocks db and set to 128MB

ptschip committed:
Increase max undo file size to 64MB

ptschip committed:
Create a database key that will sort by time for the Undo files

Do the same as we did for storing blocks to level db. This key
will sort the undo files on disk in sequential order and will prevent
excessive database compactions from happening. As a result performance
is much improved during sync.

ptschip commented on src/blockdb/blockdb.h:
```diff
@@ -0,0 +1,118 @@
+    {
+        assert(_block.IsNull() == false);
+        this->block = _block;
+        this->blockVersion = this->block.nVersion;
```
Not sure I see the need for a BlockDBValue.  I see that you're trying to store blockheight and version but that data is already in the blockindex?  The main problem though I see here is that to create this blockdbvalue you have to copy the block each time to create it.

Greg-Griffith replied:
it was outlined in the original spec zerg wrote so I implemented it as such. you make a good point though, so I will ask him about it.
reference: https://bitco.in/forum/threads/devpool-work-replace-block-files-with-key-value-db.8532/

ptschip commented:
Ok I looked, it doesn't say anything about why it's needed in the spec
so I'm still wondering what @gandrewstone was intending.  If it wasn't a
perf hit I wouldn't care that much but since we're copying blocks (which
will get big one day) I think this needs some careful study.


On 25/06/2018 9:16 AM, Greg Griffith wrote:
>
> *@Greg-Griffith* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/blockdb/blockdb.h
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/989#discussion_r197855620>:
>
> > +struct BlockDBValue
> +{
> +    int32_t blockVersion;
> +    uint64_t blockHeight;
> +    CBlock block;
> +
> +    BlockDBValue()
> +    {
> +        SetNull();
> +    }
> +
> +    BlockDBValue(const CBlock &_block)
> +    {
> +        assert(_block.IsNull() == false);
> +        this->block = _block;
> +        this->blockVersion = this->block.nVersion;
>
> it was outlined in the original spec zerg wrote.
> https://bitco.in/forum/threads/devpool-work-replace-block-files-with-key-value-db.8532/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/989#discussion_r197855620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0CMMmhhiad5UOtAy14-GSF7ElVHDks5uAQzMgaJpZM4SaqAE>.
>

ptschip committed:
Use already existing params to calculate max_file_size and write_buffer_size

Commit:
add readblock and readundo functions for leveldb

add readblock and readundo functions so we can deserialize the values of cblock and cblockundo directly into the object we want instead of deserializing them to a struct and then copying from the struct to the object we want to store them in.

Commit:
fix time inconsistencies with blockundo I/O

Commit:
add missing cases, change remaining keys to new format

Commit:
remove blockheight from blockdbvalue struct

Commit:
add missing blocktree param to test and merge dev

ptschip commented on src/blockdb/blockdb.cpp:
```diff
@@ -48,12 +48,23 @@ bool ReadBlockFromDB(const CBlockIndex *pindex, BlockDBValue &value, CBlock &blo
+    {
+        hashBlock = pindex->GetBlockHash();
+        pindex->GetBlockTime();
+    }
```
pindex->GetBlockTime() not assigned to anything...do you need it here?

ptschip commented on src/blockdb/blockdb.cpp:
```diff
@@ -64,12 +75,23 @@ bool UndoWriteToDB(const CBlockUndo &blockundo, const uint256 &hashBlock, const
+    {
+        hashBlock = pindex->GetBlockHash();
+        pindex->GetBlockTime();
+    }
```
same - pindex->GetBlockTime()

Greg-Griffith replied:
good catches. there is supposed to be an assignment there.

Greg-Griffith commented:
notes from dev meeting on 7/3:
- remove blockdbvalue struct, just read/write directly to block.
- remove readblock function since it is no longer needed

Commit:
remove blockdbvalue struct, read/write block directly

Commit:
check both blocktrees when trying a repair

Commit:
Improve sync logic so we don't end with a partially broken block index, this inevitably still happens sometimes, so run a repair function after we sync storage just in case.

ptschip committed:
Output a UI message to indicate the block database is being upgraded.

Also, print out a LOG message only when we are actually upgrading
the database. And ,we should always output the log message not just when
PRUNE logging is set.

ptschip committed:
Fix the downgrade process from blockdb to sequential files

Was not using pindex->pprev when looking up undo files.

Also the nStatus was not getting removed when a block or undo file
didn't exist or couldn't get read from the blockdb. This was causing
us to set every header as though it had block data and undo files
when in fact it didn't yet. As a result we we're not able to continue
with a sync if it hadn't yet finished.

ptschip committed:
Add blockdb files to .formatted-files and format them

ptschip committed:
Tidy up comments and remove unnecessary ones.

Also add a few empty lines to break up the code sections and
make them easier to read.

ptschip committed:
FlushStateToDisk() after we've finished a sync.

ptschip committed:
Remove section on repariing the blockindex after a sequential_files downgrade

Add else if

remove triple backslash

ptschip committed:
Clarify the log output message when block database upgrade has finished

Commit:
fix CBlockTreeDB to use nullptr not null.

Commit:
remove files that are no longer needed when changing storage modes

Commit:
set default storage mode to sequential files

ptschip commented on src/blockdb/wrapper.cpp:
```diff
@@ -201,6 +203,27 @@ void SyncStorage(const CChainParams &chainparams)
+                pblockdb->WriteBatch(batch);
+                // you must use NULL here, not nullptr
+                CBlockIndex* indexfront = blocksToRemove.front();
+                std::ostringstream frontkey;
```
when using a pointer we prefix with a "p"   , so pindexfront .  vector is v, so vBlocksToRemove, set , is set, map is map etc., number is n.

ptschip commented on src/blockdb/wrapper.cpp:
```diff
@@ -214,10 +237,26 @@ void SyncStorage(const CChainParams &chainparams)
         {
+            bool needData = true;
+            bool needUndo = true;
             CBlockIndex *index;
```
i don't see that these flags (needData needUndo) ever gets modified anywhere,  it's always true from what I can see...do you really need it for something?

Commit:
fix BLOCK_DB_MODE set logic and cleanup some comments

Commit:
add a vector to hold pairings of block hashes to their size so we can prune the db based on this information

Commit:
when a block is pruned in db mode, remove status flags and add to dirty index

ptschip commented:
Also you need some sort of python test to validate with blockd db on and then off and doing the upgrade/ downgrade

Commit:
formatting fixes

ptschip commented:
Two bugs:  the downgrade to sequential doesn't work, it gets through most of it but then complains at the end about db corruption...probably when the blockindex is written?  Also when you launch in sequential mode for the first time (no database folders created yet)...it also fails with complaining about db corruption.

Observations about pruning which you probably know about, is that it's painfully slow...about 3 times slower to do a sync.

ptschip commented:
The travis run is failing because of formatting issues....

gandrewstone commented on src/chain.h:
```diff
@@ -370,6 +371,8 @@ class CDiskBlockIndex : public CBlockIndex
     }
 
+    friend bool operator<(const CDiskBlockIndex &a, const CDiskBlockIndex &b) { return a.nHeight < b.nHeight; }
+    friend bool operator>(const CDiskBlockIndex &a, const CDiskBlockIndex &b) { return a.nHeight > b.nHeight; }
```
How are these used?  Is their use consistent with the possibility that nHeight might be the same (orphan block)

Greg-Griffith replied:
we only use this to run std::sort() so we can sort the indexes with some accuracy when we prepare to upgrade/downgrade the blockdb. we only need the < operator for this, > will be removed as it is extra

gandrewstone commented on src/dbwrapper.h:
```diff
@@ -184,7 +199,7 @@ class CDBWrapper
 {
     friend const std::vector<unsigned char> &dbwrapper_private::GetObfuscateKey(const CDBWrapper &w);
 
-private:
```
use protected here so that CBlockDB can access these fields.

gandrewstone commented on src/main.cpp:
```diff
@@ -220,10 +212,24 @@ std::unique_ptr<CRollingBloomFilter> txn_recently_in_block GUARDED_BY(cs_recentR
 std::set<int> setDirtyFileInfo GUARDED_BY(cs_main);
 
-} // anon namespace
+std::vector<CBlockFileInfo> vinfoBlockFile;
```
Let's get all these globals into globals.cpp to ensure their proper construction order.  Esp. important for any std:: or guarded data structures to ensure that they aren't destructed after their guard, for example.

gandrewstone commented on src/main.cpp:
```diff
@@ -2188,18 +2051,13 @@ bool ConnectBlock(const CBlock &block,
-                prevHash = pindex->pprev->GetBlockHash();
-            else
-                prevHash.SetNull();
-            if (!UndoWriteToDisk(blockundo, _pos, prevHash, chainparams.MessageStart()))
```
reviewer's note: the removed logic was moved into UndoWriteToDisk

gandrewstone commented on src/blockdb/wrapper.h:
```diff
@@ -0,0 +1,70 @@
+    CDiskBlockPos &pos,
+    const CBlockIndex *pindex,
+    const CMessageHeader::MessageStartChars &messageStart);
+bool UndoReadFromDisk(CBlockUndo &blockundo, const CDiskBlockPos &pos, const CBlockIndex *pindex);
```
These APIs have been nicely isolated and constitute the interface into block storage systems.  I would like to see them carefully inline documented, including descriptions of all parameters, and return values.  

Where it makes sense, also change the parameter names for clarity.  In particular, UndoWriteToDisk's "pindex" parameter is not a pointer to the CBlockIndex whose undo information is being written, but a pointer to the previous block.

Also, I think that this API writes/reads Undo information to the disk.  It does not undo a write to disk.  Let's call them WriteUndoToDisk/ReadUndoFromDisk.  This is consistent with WriteBlockToDisk/ReadBlockFromDisk. 

Yes, I see that they were poorly named prior to this PR, but let's fix now since they are not hidden in "main" but now APIs.

gandrewstone commented on src/blockdb/wrapper.h:
```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
```
Wrapper is overused, and it doesn't explain the most important thing, which is what's being wrapped.

Let's call this file "blockstorage.h", unless you have a better idea.

Also, consider blockdb.* -> blockleveldb.*  The naming now indicates that its a particular implementation of the block database on top of leveldb.  This extends naturally.  For example, we could add blocksqldb.* if we wanted to add a sql database.

gandrewstone commented on src/init.cpp:
```diff
@@ -1043,9 +1065,46 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
 
                 uiInterface.InitMessage(_("Opening Block database..."));
-                pblocktree = new CBlockTreeDB(nBlockTreeDBCache, false, fReindex);
+                if (BLOCK_DB_MODE == DB_BLOCK_STORAGE)
```
Let's add an "InitializeBlockStorage" API to wrapper.h, and impl in the wrapper.c.  In DB_BLOCK_STORAGE mode it does all that's below.  In SEQUENTIAL mode it does the code from lines 1005 to approx 1037.  This isolates and abstracts the block storage initialization logic.

Greg-Griffith replied:
the block file format that the upgrade code on lines 1005 to 1037 addresses havent been used in years, isn't it time to remove this all together?

gandrewstone commented on src/main.cpp:
```diff
@@ -1560,23 +1448,22 @@ static DisconnectResult DisconnectBlock(const CBlock &block, const CBlockIndex *
     CDiskBlockPos pos = pindex->GetUndoPos();
-    if (pos.IsNull())
+    // blockdb mode does not use the file pos system
+    if (pos.IsNull() && BLOCK_DB_MODE == SEQUENTIAL_BLOCK_FILES)
```
put this detail inside UndoReadFromDisk

gandrewstone commented on src/main.cpp:
```diff
@@ -3371,8 +3070,12 @@ bool ReceivedBlockTransactions(const CBlock &block,
+
     if (block.fExcessive)
+    {
         pindexNew->nStatus |= BLOCK_EXCESSIVE;
```
note that single line "if" statements are allowed in our code base.

gandrewstone commented on src/blockdb/blockdb.cpp:
```diff
@@ -0,0 +1,140 @@
+uint64_t FindFilesToPruneLevelDB(uint64_t nLastBlockWeCanPrune)
+{
+    LOG(PRUNE, "We have data for %u indexes \n", vDbBlockSizes.size());
+    uint64_t prunedCount;
```
initialize to 0

gandrewstone committed:
add leveldb compaction workaround, compact the undodb, and initialize a local

Commit:
remove block size data array and use nDataPos to store the block size instead(blockdb mode only). add pruneinterval to determine how much data we download between pruning attempts (blockdb mode only)

awemany reviewed (changes requested):
This commit introduces lots of new stuff, I think `thezerg nit picks` isn't an appropriate naming or description. I guess you're still going to squash and rebase this?

awemany commented on src/blockstorage/blockleveldb.cpp:
```diff
@@ -0,0 +1,145 @@
+
+bool WriteBlockToDB(const CBlock &block)
+{
+    // Create a key which will sort the database by the blocktime.  This is needed to prevent unnecessary
```
You use this key generation multiple times (and copied the comment). I'd replace those with an `inline` function to generate the key.

awemany commented on src/blockstorage/blockleveldb.cpp:
```diff
@@ -0,0 +1,145 @@
+    }
+    key << nBlockTime << ":" << hashBlock.ToString();
+
+    // calculate & write checksum
```
It seems that this checksum calculation can be factored out as well. (Yes, I guess the old code contained some duplication in this regard as well...)

awemany commented on src/blockstorage/blockleveldb.h:
```diff
@@ -0,0 +1,120 @@
+    {
+        s >> FLATDATA(hashChecksum);
+        s >> FLATDATA(hashBlock);
+        s >> _block;
```
After `Unserialize(..)` is done here, `blockundo` is still a `nullptr`. 

Would it make sense, for consistency reasons, to deal with `blockundo` here as well?

Greg-Griffith replied:
in the case of deserialization we don't use the blockundo pointer in the struct. this is intentional for performance reasons.

awemany replied:
Hmm, if it is for performance, can't you then simply init blockundo to what you hand over in Unserialize? That would still make it symmetric.

awemany commented on src/init.cpp:
```diff
@@ -1047,9 +1071,13 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
                 delete pcoinsdbview;
                 delete pcoinscatcher;
                 delete pblocktree;
+                delete pblocktreeother;
```
I think there should be a proper description somewhere
 of the variables `pblocktree` and `pblocktreeother` and why they exist and are named like this.

awemany commented on src/blockstorage/sequential_files.cpp:
```diff
@@ -0,0 +1,301 @@
+        ((int64_t)nPruneTarget - (int64_t)nCurrentUsage) / 1024 / 1024, nLastBlockWeCanPrune, count);
+}
+
+bool WriteUndoToDiskSequenatial(const CBlockUndo &blockundo,
```
Typo in the function name: `Sequenatial`

Greg-Griffith replied:
Thanks, fixed

awemany commented on src/blockstorage/sequential_files.cpp:
```diff
@@ -0,0 +1,301 @@
+    fileout << blockundo;
+
+    // calculate & write checksum
+    CHashWriter hasher(SER_GETHASH, PROTOCOL_VERSION);
```
Again, seems redundant with the above.

awemany commented on src/dbwrapper.cpp:
```diff
@@ -54,19 +54,45 @@ static leveldb::Options GetOptions(size_t nCacheSize)
 }
 
-CDBWrapper::CDBWrapper(const fs::path &path, size_t nCacheSize, bool fMemory, bool fWipe, bool obfuscate)
+static void OverrideOptions(leveldb::Options &_options, COverrideOptions *_pOverride)
```
Speaking of commits and squashing: I think this ability to specify override options is something that might be interesting in its own commit.

awemany commented on src/globals.cpp:
```diff
@@ -72,6 +72,69 @@ CCriticalSection cs_main;
 BlockMap mapBlockIndex GUARDED_BY(cs_main);
 CChain chainActive GUARDED_BY(cs_main);
 std::map<NodeId, CNodeState> mapNodeState GUARDED_BY(cs_main); // nodestate.h
+// BU variables moved to globals.cpp
```
If you can split this into individual commits, moving of the variables is in general pretty independent of your blockdb patch and should live in its own commit(s) as well IMO.

awemany commented on src/rpc/blockchain.cpp:
```diff
@@ -1056,6 +1060,50 @@ UniValue getmempoolinfo(const UniValue &params, bool fHelp)
     return mempoolInfoToJSON();
 }
 
+UniValue setpruneinterval(const UniValue &params, bool fHelp)
```
If it is a `CTweak`, isn't it already accessible using `get` and `set` and this and the other RPC therefore redundant?

awemany commented on src/txdb.cpp:
```diff
@@ -64,6 +69,25 @@ CCoinsViewDB::CCoinsViewDB(size_t nCacheSize, bool fMemory, bool fWipe)
+{
+    LOCK(cs_utxo);
+    uint256 hashBestChain;
+    if (BLOCK_DB_MODE == SEQUENTIAL_BLOCK_FILES)
```
Code like this would become superfluous if you'd have a single "CBlocksDatabase" (or such) object that you can swap out.

awemany commented on src/txdb.cpp:
```diff
@@ -72,6 +96,33 @@ uint256 CCoinsViewDB::GetBestBlock() const
+{
+    LOCK(cs_utxo);
+    uint256 hashBestChain;
+    if (!db.Read(DB_BEST_BLOCK_BLOCKDB, hashBestChain))
```
If the `db.Read(..)` fails, will that touch `hashBestChain` at all? If not, this code seems, except for the locking, quite cumbersome, doesn't it? (Sorry, I am by far not a `leveldb` expert...)

Greg-Griffith replied:
the code you are questioning here is much older than this pr, im unsure what the original reasons for this were

awemany replied:
I see, thanks. I wonder whether it could be cleaned up (later).

awemany reviewed (changes requested):
This is all just from browsing the code, I need to play with this to give better feedback.

As a general comment, what I am missing is some kind of sense of a clean interface where I can plug in either the old file-based DB or the new leveldb. This is probably too much to ask for for this PR, but can the code be cleaned up in that sense?

Then, the mixing of moving of code with the implementation of the new `blocksdb` approach makes it very hard to follow the changes. If you have any change to split this up into moves and into additions for the reader, I would very much appreciate that :-)

awemany commented on src/init.cpp:
```diff
@@ -1070,7 +1070,15 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
+                 */
                 delete pblocktree;
+                /**
+                 * pblocktreeother provices access to some other block metadata database
```
I have no clue what 'some other block metadata database' is, TBH :D
I am still confused by this.

Commit:
move globals to globals.cpp, remove unused CDiskBlockIndex operator, change public members of CDBWrapper to protected

Commit:
Rename blockdb folder to blockstorage. rename blockdb/wrapper.* to blockstorage/blockstorage.*

Commit:
move pblocktree pointer initialization to InitializeBlockStorage api in blockstorage/blockstorage.cpp,
fix pointers to use nullptr instead of NULL

awemany commented on src/blockdb/wrapper.cpp:
```diff
@@ -200,18 +278,48 @@ void SyncStorage(const CChainParams &chainparams)
+                pindexNew->nFile =          tempindex->nFile;
+                pindexNew->nDataPos =       tempindex->nDataPos;
+                pindexNew->nUndoPos =       tempindex->nUndoPos;
+                pindexNew->nVersion =       tempindex->nVersion;
```
This series of assignments and some other stuff seems to be redundant with the other implementation above?

awemany commented on src/txdb.cpp:
```diff
@@ -333,20 +330,20 @@ bool CBlockTreeDB::FindBlockIndex(uint256 blockhash, CBlockIndex& pindex)
+                    pindex->nDataPos = diskindex.nDataPos;
+                    pindex->nUndoPos = diskindex.nUndoPos;
+                    pindex->nVersion = diskindex.nVersion;
+                    pindex->hashMerkleRoot = diskindex.hashMerkleRoot;
```
same here ..

awemany commented on src/blockdb/blockdb_leveldb.cpp:
```diff
@@ -0,0 +1,106 @@
+bool CFullBlockDB::WriteBatchSync(const std::vector<CBlock> &blocks)
+{
+    CDBBatch batch(*this);
+    for (std::vector<CBlock>::const_iterator it = blocks.begin(); it != blocks.end(); it++)
```
nit: `for (auto ...` or `for (<type> ... : ...)` looks nicer

awemany commented on src/txdb.cpp:
```diff
@@ -260,6 +283,60 @@ bool CBlockTreeDB::ReadFlag(const std::string &name, bool &fValue)
+                    pindex.nHeight = diskindex.nHeight;
+                    pindex.nFile = diskindex.nFile;
+                    pindex.nDataPos = diskindex.nDataPos;
+                    pindex.nUndoPos = diskindex.nUndoPos;
```
again, redundancy here..

awemany commented on src/blockdb/blockdb.cpp:
```diff
@@ -35,6 +35,35 @@ bool ReadBlockFromDB(const CBlockIndex *pindex, BlockDBValue &value)
+bool UndoWriteToDB(const CBlockUndo &blockundo, const uint256 &hashBlock)
+{
+    // calculate & write checksum
+    CHashWriter hasher(SER_GETHASH, PROTOCOL_VERSION);
```
nit: You have this pattern so often in your code that this IMO asks for some further code reduction with a varargs function that hashes a string of objects using the default network parameters.

awemany replied:
Also (but maybe that is an idea for later), can't the CBlockUndo be extended with the hash of the block?

gandrewstone commented on src/init.cpp:
```diff
@@ -746,6 +747,7 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
         LOGA("Prune configured to target %uMiB on disk for block and undo files.\n", nPruneTarget / 1024 / 1024);
         fPruneMode = true;
     }
+    pruneInterval = GetArg("-pruneinterval", DEFAULT_PRUNE_INTERVAL) * 1024 * 1024;
```
the CTweak will automatically initialize itself, so you do not need this line.  Also, note that doing "* 1024 * 1024" will break things because the user can change this value at any time via RPC calls.  By multiplying it by 1 MiB here, you make pruneinterval the number of MiB if defined in bitcoin.conf or as an argument, but its units are BYTES if it is changed by RPC call, since it does not get multiplied in that set logic.

In summary, you need to multiply pruneInterval by 1 MiB at the point where it is used.

Commit:
fix improper implementation of pruneIntervalTweak

Commit:
Flush index to disk when pruning leveldb, write to leveldb using fsync

Flush index to disk when pruning leveldb before removing block from db 
to ensure we dont have a state corruption when program quits 
unexpectedly. 
use fsync when writing to leveldb to make sure we dont return out of the 
write function until data is pushed all the way to persistent storage

Commit:
wrap folder removal in a try catch block due to boost::fs::remove_all behaving differently on windows compared to unix based systems.

Commit:
fix pruneIntervalTweak for value vs Value(), formatting fixes

ptschip commented:
looks like a failure in one of the python tests...

gandrewstone commented:
VM seems to have just been halted in the middle of the wallet.py test.  Restarting
